drivers/floorsense/lockers_websocket.cr

```diff
@@ -0,0 +1,977 @@
+  end
+
+  # a list of lockers that are allocated to the user
+  @[Security(Level::Administrator)]
+  def lockers_allocated_to(user_id : String) : Array(PlaceLocker)
+    lockers = all_lockers
+      .map do |locker_info|
+        if user_id == locker_info.uid
+          PlaceLocker.new(locker_info)
+        end
+      end
+      .compact
+  end
+
+  @[Security(Level::Administrator)]
+  def locker_share(
+    # Used as reservation_id
+    bank_id : String | Int64,
+    locker_id : String | Int64,
+    owner_id : String,
+    # Used as user_id
+    share_with : String
+  ) : Nil
+    locker_share(bank_id.to_s, share_with.to_s)
+  end
+
+  @[Security(Level::Administrator)]
+  def locker_unshare(
+    # Used as reservation_id
+    bank_id : String | Int64,
+    # Used as user_id
+    locker_id : String | Int64,
+    owner_id : String,
+    # the individual you previously shared with
+    shared_with_id : String? = nil
+  ) : Nil
+    locker_unshare(bank_id.to_s, locker_id.to_s)
+  end
+
+  # a list of user-ids that the locker is shared with.
+  # this can be placeos user ids or emails
+  @[Security(Level::Administrator)]
+  def locker_shared_with(
+    # Used as reservation_id
+    bank_id : String | Int64,
+    locker_id : String | Int64,
+    owner_id : String
+  ) : Array(String)
+    locker_shared?(bank_id.to_s).map do |shared_with|
+      shared_with.as_h.["uid"].to_s
+    end
+  end
+
+  @[Security(Level::Administrator)]
+  def locker_unlock(
+    bank_id : String | Int64,
+    locker_id : String | Int64,
+
+    # sometimes required by locker systems
+    owner_id : String? = nil,
+    # time in seconds the locker should be unlocked
+    # (can be ignored if not implemented)
+    open_time : Int32 = 60,
+    # optional pin code - if user entered from a kiosk
+    pin_code : String? = nil
+  ) : Nil
+    _locker_unlock(locker_id.to_s, owner_id.to_s)
+  end
+
+  # ========================================
+  # Locatable Interface
+  # ========================================
+
+  # array of devices and their x, y coordinates, that are associated with this user
+  def locate_user(email : String? = nil, username : String? = nil)
+    logger.debug { "smartalock is incapable of locating users" }
+    [] of Nil
+  end
+
+  # return an array of MAC address strings
+  # lowercase with no seperation characters abcdeffd1234 etc
+  def macs_assigned_to(email : String? = nil, username : String? = nil) : Array(String)
+    logger.debug { "smartalockis incapable of listing macs assigned to an user or an email" }
+    [] of String
+  end
+
+  # return `nil` or `{"location": "wireless", "assigned_to": "bob123", "mac_address": "abcd"}`
+  def check_ownership_of(mac_address : String) : OwnershipMAC?
+    logger.debug { "smartalock is incapable of checking ownerships" }
+    nil
+  end
+
+  def device_locations(zone_id : String, location : String? = nil)
+    logger.debug { "smartalock is incapable to locate devices" }
+    [] of Nil
+  end
```

Comment:
do we need this code as it no longer implements the locatable interface

Reply:
Seems like it was required by the actual interface.

Reply:
ahh yes! Could you please implement the `device_locations` code similar to what is happening on releezme

Needs to output data like this:
https://docs.placeos.com/how-to/location-services/location-services#lockers

Reply:
I don't think it supports such a case, I will investigate.

Reply:
effectively this function needs to grab all the currently active locker bookings (like bookings that are live right now, excluding past and future bookings) and filter by level or building zones

so the zone_ids need to be converted - basically very similar to [this code, just for lockers](https://github.com/PlaceOS/drivers/blob/master/drivers/floorsense/location_service.cr#L154-L221).

So you'll need to [maintain mappings, like this code](https://github.com/PlaceOS/drivers/blob/master/drivers/floorsense/location_service.cr#L19-L37)

I still think you should have followed the recommendation from the PPT to keep the locker location service separate from the drivers code (basically just create a `locker_location_service.cr`) that's the same as `location_service.cr` just for lockers
![image](https://github.com/PlaceOS/drivers/assets/368013/5abd53f3-b701-4bd3-82c8-25d58efeeea2)

as this would have allowed the settings in `location_service.cr` to be shared with `locker_location_service.cr` - only one place where the zone_ids => floorsense id mappings need to be maintained

Reply:
I did see that comment but it was a bit late, thank you, will do!